SPARK-11326: Split networking in standalone mode

jacek-lewandowski opened the pull request.

jacek-lewandowski commented:
jenkins, test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

jacek-lewandowski commented:
@vanzin could you take a look?

SparkQA commented:
**[Test build #44379 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44379/consoleFull)** for PR 9287 at commit [`f3b2d9f`](https://github.com/apache/spark/commit/f3b2d9f51b1b28da9e74646024d9fb7ec4a6df9d).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44380 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44380/consoleFull)** for PR 9287 at commit [`f3b2d9f`](https://github.com/apache/spark/commit/f3b2d9f51b1b28da9e74646024d9fb7ec4a6df9d).

SparkQA commented:
**[Test build #44379 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44379/consoleFull)** for PR 9287 at commit [`f3b2d9f`](https://github.com/apache/spark/commit/f3b2d9f51b1b28da9e74646024d9fb7ec4a6df9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44379/
Test PASSed.

SparkQA commented:
**[Test build #44380 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44380/consoleFull)** for PR 9287 at commit [`f3b2d9f`](https://github.com/apache/spark/commit/f3b2d9f51b1b28da9e74646024d9fb7ec4a6df9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44380/
Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

jacek-lewandowski commented:
jenkins, test this please

SparkQA commented:
**[Test build #44594 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44594/consoleFull)** for PR 9287 at commit [`281d4f4`](https://github.com/apache/spark/commit/281d4f44f7c237a8a76db47ea61a4c981a28a409).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44594 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44594/consoleFull)** for PR 9287 at commit [`281d4f4`](https://github.com/apache/spark/commit/281d4f44f7c237a8a76db47ea61a4c981a28a409).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44594/
Test FAILed.

SparkQA commented:
**[Test build #44596 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44596/consoleFull)** for PR 9287 at commit [`281d4f4`](https://github.com/apache/spark/commit/281d4f44f7c237a8a76db47ea61a4c981a28a409).

SparkQA commented:
**[Test build #44596 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44596/consoleFull)** for PR 9287 at commit [`281d4f4`](https://github.com/apache/spark/commit/281d4f44f7c237a8a76db47ea61a4c981a28a409).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44596/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44657 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44657/consoleFull)** for PR 9287 at commit [`62bf841`](https://github.com/apache/spark/commit/62bf8415eaddd3260bcef7f0a74e7963f5bdb250).

SparkQA commented:
**[Test build #44657 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44657/consoleFull)** for PR 9287 at commit [`62bf841`](https://github.com/apache/spark/commit/62bf8415eaddd3260bcef7f0a74e7963f5bdb250).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44657/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44999 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44999/consoleFull)** for PR 9287 at commit [`84baafc`](https://github.com/apache/spark/commit/84baafc0a09504435e0c6058f6d1e607dae2fdab).

SparkQA commented:
**[Test build #44999 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44999/consoleFull)** for PR 9287 at commit [`84baafc`](https://github.com/apache/spark/commit/84baafc0a09504435e0c6058f6d1e607dae2fdab).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44999/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45119 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45119/consoleFull)** for PR 9287 at commit [`3c1467a`](https://github.com/apache/spark/commit/3c1467ab94f5befa0c9a6c0f6b8aaaa4c1701002).

SparkQA commented:
**[Test build #45119 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45119/consoleFull)** for PR 9287 at commit [`3c1467a`](https://github.com/apache/spark/commit/3c1467ab94f5befa0c9a6c0f6b8aaaa4c1701002).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n

vanzin commented on core/src/main/scala/org/apache/spark/Namespace.scala:
```diff
@@ -0,0 +1,83 @@
+  /** Namespace name for conf property */
+  def confNamespace: String
+
+  /** Adds this namespace to the given env variable name. Assuming the given `envKey` is
```
this doesn't follow the multi-line scaladoc format used in Spark.

    /**
     * Text.
     */

Same in other places.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45119/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/SecurityManager.scala:
```diff
@@ -483,7 +482,7 @@ private[spark] object SecurityManager {
   // This is used to set auth secret to an executor's env variable. It should have the same
   // value as SPARK_AUTH_SECRET_CONF set in SparkConf
-  val ENV_AUTH_SECRET = "_SPARK_AUTH_SECRET"
+  val ENV_AUTH_SECRET = "SPARK_AUTH_SECRET"
```
Why this change? The `_` is normally used to indicate something that users should not set directly, and that's the case with this env variable.

vanzin commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -364,6 +364,12 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
 
+  /** Returns a copy of this configuration with all the properties at `spark.srcNamespace.xxx`
+    * copied to `spark.xxx` */
+  def fromNamespace(namespace: Namespace): SparkConf = {
```
`Namespace` is `private[spark]` but this method is public. Scala should have printed a warning about this. But that means this method should be `private[spark]`.

vanzin commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -364,6 +364,12 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
   /** Does the configuration contain a given parameter? */
   def contains(key: String): Boolean = settings.containsKey(key)
 
+  /** Returns a copy of this configuration with all the properties at `spark.srcNamespace.xxx`
```
nit: param name is `namespace` not `srcNamespace`.

vanzin commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -44,6 +44,7 @@ import org.apache.hadoop.mapreduce.{InputFormat => NewInputFormat, Job => NewHad
 import org.apache.hadoop.mapreduce.lib.input.{FileInputFormat => NewFileInputFormat}
 
 import org.apache.mesos.MesosNativeLibrary
+import org.apache.spark.SecurityManager._
```
nit: import order

vanzin commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -379,6 +380,16 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
+
+    // if a secret key has been set in an env variable, use it
+    for (ns <- Seq(Namespace.Blank, Namespace.Application, Namespace.Submission)) {
+      sys.env.get(ns.inEnv(ENV_AUTH_SECRET))
```
Now that I see the methods being used, the names sound kinda weird. I'd prefer `ns.envName` and `ns.confName`.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/Client.scala:
```diff
@@ -23,9 +23,10 @@ import scala.reflect.ClassTag
 
 import org.apache.spark.rpc.{RpcEndpointRef, RpcAddress, RpcEnv, ThreadSafeRpcEndpoint}
-import org.apache.spark.{Logging, SecurityManager, SparkConf}
+import org.apache.spark.{Namespace, Logging, SecurityManager, SparkConf}
```
nit: import order

vanzin commented on core/src/main/scala/org/apache/spark/deploy/Client.scala:
```diff
@@ -23,9 +23,10 @@ import scala.reflect.ClassTag
 import scala.util.{Failure, Success}
 
 import org.apache.log4j.{Level, Logger}
+import org.apache.spark.SecurityManager.{SPARK_AUTH_SECRET_CONF, ENV_AUTH_SECRET}
```
nit: import order

jacek-lewandowski replied:
Weird that IDE didn't take care of that :(

vanzin commented on core/src/main/scala/org/apache/spark/deploy/Client.scala:
```diff
@@ -80,11 +81,30 @@ private class ClientEndpoint(
+
+        // we want to transport secrets only from submission and app namespaces; they will be
+        // transported as environment variables though
+        val envSecrets = Set(Namespace.Submission, Namespace.Application).flatMap { ns =>
```
I think used `Seq` in these places would be better, since it's faster and you don't really need the functionality of a `Set` here.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/Client.scala:
```diff
@@ -80,11 +81,30 @@ private class ClientEndpoint(
+          mainClass = mainClass,
+          arguments = Seq("{{WORKER_URL}}", "{{USER_JAR}}", driverArgs.mainClass)
+            ++ driverArgs.driverOptions,
+          environment = sys.env ++ envSecrets - ENV_AUTH_SECRET,
```
what's the precedence order here? can you add parentheses to make that explicit (and make sure the code does what's intended)?

vanzin commented on core/src/main/scala/org/apache/spark/deploy/Client.scala:
```diff
@@ -208,6 +228,9 @@ private class ClientEndpoint(
  * Executable utility for starting and terminating drivers inside of a standalone cluster.
  */
 object Client {
+  val SYSTEM_NAME = "driverClient"
```
these should not be public (since `Client` is a public class).

vanzin commented on core/src/main/scala/org/apache/spark/deploy/Client.scala:
```diff
@@ -226,12 +249,29 @@ object Client {
-    val rpcEnv =
-      RpcEnv.create("driverClient", Utils.localHostName(), 0, conf, new SecurityManager(conf))
+    // if a secret key has been set in an env variable, use it
+    sys.env.get(ENV_AUTH_SECRET).foreach(conf.set(SPARK_AUTH_SECRET_CONF, _))
```
This should never happen. The user is not supposed to set that env variable.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/Client.scala:
```diff
@@ -226,12 +249,29 @@ object Client {
+    sys.env.get(ENV_AUTH_SECRET).foreach(conf.set(SPARK_AUTH_SECRET_CONF, _))
+
+    // if a secret key for submission has been set in an env variable, use it
+    sys.env.get(Namespace.Submission.inEnv(ENV_AUTH_SECRET))
```
As above, this code should not be added.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala:
```diff
@@ -17,17 +17,16 @@
+import org.apache.spark.deploy.{ApplicationDescription, ExecutorState}
 import org.apache.spark.rpc._
 import org.apache.spark.util.{RpcUtils, ThreadUtils, Utils}
+import org.apache.spark.{Logging, SecurityManager, SparkConf, SparkContext, Namespace}
```
nit: import order.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala:
```diff
@@ -99,8 +98,10 @@ private[spark] class AppClient(
             logInfo("Connecting to master " + masterAddress.toSparkURL + "...")
-            val masterRef =
-              rpcEnv.setupEndpointRef(Master.SYSTEM_NAME, masterAddress, Master.ENDPOINT_NAME)
+            val masterRef = rpcEnv.setupEndpointRef(
```
I see changes like this in lots of places. These are not arguments with default values, so calling them like this is completely unnecessary and just adds noise. Can you please revert changes like this from your patch?

jacek-lewandowski commented:
@vanzin thanks a lot for your review. Of course, I'll apply all those rename/import/style changes. But the thing about env variable needs explanation - actually the user is supposed to provide a secret key in env variable. Especially, it is much more secure than providing it in a command line of spark-submit (as you know). Now that the network communication layers are separated, the user would have to provide "credentials" in order to submit the application.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala:
```diff
@@ -303,3 +305,25 @@ private[spark] class AppClient(
 
 }
+
+object AppClient {
```
This should not be public.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -30,7 +30,7 @@ import scala.util.Random
 
 import org.apache.spark.rpc._
-import org.apache.spark.{Logging, SecurityManager, SparkConf, SparkException}
+import org.apache.spark.{Logging, SecurityManager, SparkConf, SparkException, Namespace}
```
nit: import order

vanzin commented:
> But the thing about env variable needs explanation - actually the user is supposed to provide a secret key in env variable.

I haven't finished my review yet, but I was going to get to that. That change is really not necessary.

For the secret needed to connect to the Spark Master, that can be in the configuration file. Everybody needs to share the same secret, so there's no problem in it being there. Until the Master can do proper authentication, we'll have to live with that fact.

For the per-app secret, it should be auto-generated, like in YARN. The user should not need to provide their own secret every time they launch their app.

So you don't really have to set any env variables. The env variable is just a way to get the secret from the Worker to the executor without using the command line.

BTW, I think this is a step in the right direction, but because of the above, this is not really adding any extra security. Because all Spark-launched processes will run as the same user, it's possible for any malicious executor to just read other app's secrets from the Worker's memory (or directly from other running executors' environment in /proc).

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -47,11 +47,15 @@ import org.apache.spark.util.{ThreadUtils, SignalLogger, Utils}
     val securityMgr: SecurityManager,
     val conf: SparkConf)
-  extends ThreadSafeRpcEndpoint with Logging with LeaderElectable {
+  extends ThreadSafeRpcEndpoint with Logging with LeaderElectable { master =>
```
nit: in other Spark code, `master =>` is on its own line.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -126,27 +131,92 @@ private[deploy] class Master(
+    }
+
+    override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
+      case msg @ RequestSubmitDriver(description) =>
```
since you're not really using any of the arguments, can you just say `case msg: RequestSubmitDriver =>` here? I'm not sure Scala allows that for case classes but that would be much cleaner.

If you can't, then you can just use placeholders (`_`) for the arguments.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -126,27 +131,92 @@ private[deploy] class Master(
+
+  private var rpcSubmissionEndpointRef: Option[RpcEndpointRef] = None
+
+  object RPCSubmissionEndpoint extends ThreadSafeRpcEndpoint {
```
it's not really clear: why do you need this? why can't you just use the Master itself, since it already handles all these messages?

At the very least, this needs a comment (in the code) explaining that. Also, nested classes are generally declared at the bottom of the enclosing class.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -1051,32 +1133,61 @@ private[deploy] class Master(
   val ENDPOINT_NAME = "Master"
+  val SUBMISSION_ENDPOINT_NAME = "Master"
+
+  val DefaultRpcPort = 7077
```
nit: constants are DEFINED_IN_CAPS. Just look a few lines above.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala:
```diff
@@ -24,30 +24,35 @@ import org.apache.spark.util.{IntParam, Utils}
-  if (System.getenv("SPARK_MASTER_WEBUI_PORT") != null) {
-    webUiPort = System.getenv("SPARK_MASTER_WEBUI_PORT").toInt
-  }
+  sys.env.get("SPARK_MASTER_HOST").foreach(v => host = v)
```
nit: `foreach { v => ... }` (we use `()` only when using `_`)

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala:
```diff
@@ -24,30 +24,35 @@ import org.apache.spark.util.{IntParam, Utils}
+  sys.env.get("SPARK_MASTER_PORT").foreach(v => port = v.toInt)
+  sys.env.get("SPARK_MASTER_WEBUI_PORT").foreach(v => webUiPort = v.toInt)
+
+  sys.env.get("SPARK_MASTER_SUBMISSION_RPC_PORT").orElse(sys.env.get("SPARK_MASTER_PORT"))
```
Why add a new name for the env variable? Is `SPARK_MASTER_PORT` gonna change meaning because of this change?

Changing configuration is kinda bad because you still need to keep the old code, and need to educate users on the new values. So unless there's a very good reason for the change, I don't see the point.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala:
```diff
@@ -24,30 +24,35 @@ import org.apache.spark.util.{IntParam, Utils}
+  conf.getOption("spark.master.ui.port").foreach(v => webUiPort = v.toInt)
+  conf.getOption("spark.master.rpc.port").foreach(v => port = v.toInt)
+  conf.getOption("spark.master.submission.rest.port").foreach(v => restSubmissionPort = v.toInt)
+  conf.getOption("spark.master.submission.rpc.port").foreach(v => rpcSubmissionPort = v.toInt)
```
It's really confusing to have conf split like this. You should merge the two pieces of code setting `rpcSubmissionPort` into one statement.

I'd also recommend declaring `vals` and using `getOrElse` instead of `foreach`.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/MasterArguments.scala:
```diff
@@ -68,6 +73,14 @@ private[master] class MasterArguments(args: Array[String], conf: SparkConf) {
       webUiPort = value
       parse(tail)
 
+    case ("--rpc-submission-port") :: IntParam(value) :: tail =>
```
I'd rather not add command line parameters. Just keep everything in the conf file.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/CommandUtils.scala:
```diff
@@ -22,8 +22,8 @@ import java.io.{File, FileOutputStream, InputStream, IOException}
 
-import org.apache.spark.Logging
-import org.apache.spark.SecurityManager
+import org.apache.spark.{Namespace, Logging, SecurityManager}
```
nit: import order

vanzin commented on core/src/main/scala/org/apache/spark/deploy/worker/CommandUtils.scala:
```diff
@@ -117,4 +114,15 @@ object CommandUtils extends Logging {
     }.start()
   }
+
+  val SystemPropertyPattern = """^-D(.+?)=(.*)$""".r
```
nit: either ALL_CAPS or camelCase. We don't declare variables with names like that in Spark.

markhamstra commented:
Please see my JIRA comment.

vanzin commented:
I stopped when I got to the tests; overall, I share the same concerns Mark mentioned on the bug. I don't really know what the goals of standalone mode are in the long run, but if your concern is security, there's a lot of much harder work needed to get that working.

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
**[Test build #45217 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45217/consoleFull)** for PR 9287 at commit [`b83ab2c`](https://github.com/apache/spark/commit/b83ab2ca6c2d6d05feae1cd3e06ff4f0bcabe832).

SparkQA commented:
**[Test build #45217 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45217/consoleFull)** for PR 9287 at commit [`b83ab2c`](https://github.com/apache/spark/commit/b83ab2ca6c2d6d05feae1cd3e06ff4f0bcabe832).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45217/
Test FAILed.

vanzin commented:
BTW you probably want @andrewor14 to take a look at this stuff.

AmplabJenkins commented:
Build triggered. sha1 is original commit.

AmplabJenkins commented:
Build started sha1 is original commit.

jacek-lewandowski commented:
@vanzin I've changed it a little bit. The master doesn't create two endpoints now but different users can be used with a single endpoint. So basically the user can plug his own password authenticator if he wants to which will be used to authenticate Spark users with the master. 

To authenticate workers with the master, and clients when no special password authenticator is plugged in, the default SASL user is used (as it was used so far). When the user wants to enable more sophisticated authentication, he simply needs to provide some authenticator or just set different secrets for the default user {{sparkSaslUser}} (which is used by the workers) and other users (the default authenticator allows to set multiple secrets like {{spark.authenticate.secrets.<username>=password}} while the default password is set as usual with {{spark.authenticate.secret}}). In this case, the users should just do not need to know the default user password. Obviously, this only works for Netty RPC. The whole original behaviour when no-secret or a single shared secret is provided remains unchanged.

The last PR is not polished, but it removes most of the changes from the master. The remaining thing is authorisation - determining who is communicating with the master and authorising the requests (for example, only a default user can exchange worker messages, or only the application owner can remove his application).

SparkQA commented:
**[Test build #45230 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45230/consoleFull)** for PR 9287 at commit [`df542fc`](https://github.com/apache/spark/commit/df542fcc895fa5a25d133d0f80e139bb01aa6431).

AmplabJenkins commented:
Build finished. No test results found.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45230/

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45373 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45373/consoleFull)** for PR 9287 at commit [`f604a6d`](https://github.com/apache/spark/commit/f604a6d0651e3b206c03503c06f3d07da3846bfd).

SparkQA commented:
**[Test build #45373 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45373/consoleFull)** for PR 9287 at commit [`f604a6d`](https://github.com/apache/spark/commit/f604a6d0651e3b206c03503c06f3d07da3846bfd).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class DefaultPasswordAuthenticator(sparkConf: SparkConf)`\n  * `trait PasswordAuthenticator `\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n  * `public class ClientAddress implements Serializable `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45379 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45379/consoleFull)** for PR 9287 at commit [`508c97d`](https://github.com/apache/spark/commit/508c97da7d005303b8c9518fa84830e6724f6fa9).

SparkQA commented:
**[Test build #45379 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45379/consoleFull)** for PR 9287 at commit [`508c97d`](https://github.com/apache/spark/commit/508c97da7d005303b8c9518fa84830e6724f6fa9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class DefaultPasswordAuthenticator(sparkConf: SparkConf)`\n  * `trait PasswordAuthenticator `\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n  * `public class ClientAddress implements Serializable `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45411 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45411/consoleFull)** for PR 9287 at commit [`8b8121c`](https://github.com/apache/spark/commit/8b8121c70717b772724659be2763f716e139dd60).

SparkQA commented:
**[Test build #45411 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45411/consoleFull)** for PR 9287 at commit [`8b8121c`](https://github.com/apache/spark/commit/8b8121c70717b772724659be2763f716e139dd60).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `trait PasswordAuthenticator `\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n  * `public class ClientAddress implements Serializable `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

jacek-lewandowski commented:
jenkins, test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45420 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45420/consoleFull)** for PR 9287 at commit [`8b8121c`](https://github.com/apache/spark/commit/8b8121c70717b772724659be2763f716e139dd60).

SparkQA commented:
**[Test build #45420 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45420/consoleFull)** for PR 9287 at commit [`8b8121c`](https://github.com/apache/spark/commit/8b8121c70717b772724659be2763f716e139dd60).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `final class ShuffleSortDataFormat extends SortDataFormat<PackedRecordPointer, LongArray> `\n  * `final class UnsafeSortDataFormat extends SortDataFormat<RecordPointerAndKeyPrefix, LongArray> `\n  * `trait PasswordAuthenticator `\n  * `class ExternalShuffleService(`\n  * `  case class BoundPortsResponse(`\n  * `class ExecutionMemoryPool(`\n  * `abstract class MemoryPool(lock: Object) `\n  * `class StorageMemoryPool(lock: Object) extends MemoryPool(lock) with Logging `\n  * `abstract class Writer extends BaseReadWrite `\n  * `trait Writable `\n  * `abstract class Reader[T] extends BaseReadWrite `\n  * `trait Readable[T] `\n  * `public class ClientAddress implements Serializable `\n  * `class LDAModel(JavaModelWrapper, JavaSaveable, Loader):`\n  * `case class DropAnyNull(child: Expression) extends UnaryExpression with ExpectsInputTypes `\n  * `case class GetInternalRowField(child: Expression, ordinal: Int, dataType: DataType)`\n  * `case class Expand(`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45502 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45502/consoleFull)** for PR 9287 at commit [`fe9de83`](https://github.com/apache/spark/commit/fe9de83ec881d8c021a4f2b790dcb05efa6a8ce0).

SparkQA commented:
**[Test build #45502 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45502/consoleFull)** for PR 9287 at commit [`fe9de83`](https://github.com/apache/spark/commit/fe9de83ec881d8c021a4f2b790dcb05efa6a8ce0).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `trait PasswordAuthenticator `\n  * `class ExternalShuffleService(`\n  * `public class ClientAddress implements Serializable `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45502/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45512 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45512/consoleFull)** for PR 9287 at commit [`784cbca`](https://github.com/apache/spark/commit/784cbcaf1c1d4019f4793456d397a62ba862b342).

SparkQA commented:
**[Test build #45512 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45512/consoleFull)** for PR 9287 at commit [`784cbca`](https://github.com/apache/spark/commit/784cbcaf1c1d4019f4793456d397a62ba862b342).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `trait PasswordAuthenticator `\n  * `class ClientAddress implements Serializable `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45512/
Test PASSed.

jacek-lewandowski committed:
SPARK-11326: Separated scheduler client from application

In standalone mode, the same network endpoint was used to contact Master and to connect to executors. This commit separates those endpoint so that there the endpoint for application submission is defined at spark.submission namespace.

jacek-lewandowski committed:
SPARK-11326: Introduced customizable authenticator

The code which is responsible for obtaining a user secret was moved to DefaultPasswordAuthenticator which implements PasswordAuthenticator. The user is able to provide its own authenticator implementation.

jacek-lewandowski committed:
SPARK-11326: Changed the way how appId is passed to Sasl bootstrap

Before this change there was no way to pass either appId or username to Sasl handler. With this change, appId is kept with NettyRpcEndpointRef and is passed to Sasl each time a new connection to that endpoint is opened.

jacek-lewandowski committed:
SPARK-11326: Added authentication to external shuffle service in standalone mode

Shuffle service can use authentication in standalone mode in the same way as it is used in Yarn mode.

jacek-lewandowski committed:
SPARK-11326: Include authentication id of the connected client in RpcCallContext

jacek-lewandowski committed:
SPARK-11326: RpcEndpoint receive method gets RpcCallContext param

jacek-lewandowski committed:
SPARK-11326: Added authorization to Spark Master

Authenticated user is saved as an owner in ApplicationInfo and DriverInfo. Only that user or Spark default user can manage the application. Only Spark default user can register workers.

jacek-lewandowski committed:
SPARK-11326: Integration test suites for different authentication configs in standalone mode

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45630 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45630/consoleFull)** for PR 9287 at commit [`9f6bb88`](https://github.com/apache/spark/commit/9f6bb88cc306ea9e6a9199b5e1065d982da169c8).

jacek-lewandowski commented:
Rebased and more reasonably split into commits.

SparkQA commented:
**[Test build #45630 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45630/consoleFull)** for PR 9287 at commit [`9f6bb88`](https://github.com/apache/spark/commit/9f6bb88cc306ea9e6a9199b5e1065d982da169c8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `trait PasswordAuthenticator `\n  * `class ClientAddress implements Serializable `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45630/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45655/
Test FAILed.

jacek-lewandowski commented:
jenkins, test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45661 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45661/consoleFull)** for PR 9287 at commit [`40df523`](https://github.com/apache/spark/commit/40df523d6228888f09ea10ae5769d176a98517de).

SparkQA commented:
**[Test build #45661 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45661/consoleFull)** for PR 9287 at commit [`40df523`](https://github.com/apache/spark/commit/40df523d6228888f09ea10ae5769d176a98517de).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `trait PasswordAuthenticator `\n  * `class ClientAddress implements Serializable `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45661/
Test FAILed.

HyukjinKwon commented:
Maybe we might have to correct the title just like the others, `[SPARK-XXXX][SQL]` (this is described in https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark).
Also, it would be better if the description of this PR is updated, It is empty.

tnachen commented:
This has been stale for a while, we should close this if there is no update here.

HyukjinKwon commented:
+1 for closing.

jacek-lewandowski commented:
Closing, although I don't really understand the reason why it wasn't accepted.

HyukjinKwon commented:
I ~guess~ the reason it wasn't accepted and reviewed due to the lack of man power of committers or the lack of appealing or because it simply just got stale. At least, I said +1 for this reason.

I ~think~ it is okay to reopen, fix the codes to pass the tests, then keep pining reviewers after filling the PR description with a proper title according to the glide lines and an email to dev mailing list, yelling to the public to explain why this fix is right and nedded. Usually, it is really hard for a lot of changes to be merged quickly because it is hard to review. For this reason, I also have several old PRs but I am trying to explain and keep pinging with rebasing everytime they get conflicts

This one is not really in my area but if you strongly think this is the right fix and the important change,  I'd keep open this and then try to persuade reviwers with keeping pininging.
